### ISSUE TITLE
Please add support for 'zram' kernel module

### ISSUE THREAD

**Issue by MrHacky** [osmc/osmc#246]
This module allows creating compressed memory block devices which can then be used as swap. This effectively increases your memory at the cost of some cpu usage for (de)compression when pages need to be 'swapped'. Very useful  for low-memory devices like the raspberry pi(2). As an example, i previously had gentoo installed on my pi2, and zram enabled me to compile chromium in a timeframe measured in hours instead of days.

kernel config needed:
CONFIG_ZSMALLOC=m
CONFIG_ZRAM=m

Eventually it might even make sense to configure such a swap device by default on some devices, but for now just adding the kernel modules would be awesome so users (me :-P) can experiment with this.

**Comment by samnazarko** [comment]
Hi

I have seen regressions in performance when we tested ZRAM swaps, particularly on Raspberry Pi 1. 

While I appreciate it may help your compilation times, this is not a typical use case with OSMC. I'll check this won't cause regressions as a module, and if it doesn't, I'll schedule it for inclusion in the kernel

**Comment by MrHacky** [comment]
@samnazarko did you ever get around to test if this is acceptable for inclusion as a module?

I understand enabling this by default isn't really useful for the normal OSMC use cases, but having the module available would be really handy for me.

**Comment by samnazarko** [comment]
My understanding is if we add it as a module, it won't be enabled by default. I'm willing to add support for this and will do so in the near future.

For which application is this beneficial?

**Comment by johnfocker** [comment]
@samnazarko It is beneficial to your whole system.
For instance I am running your distro as my main media center which runs a few other apps like sonarr and couchpotato on my Pi3. The memory boost from zram would be a real plus.
Also if you're worried about performance for kodi maybe you should first reduce the default niceness of the kodi process. I've seen it increase reactivity of my osmc gui.

**Comment by samnazarko** [comment]
Hi

It's been a while since we looked at this.

Can you please provide a list of changes you would make to improve performance?

> I am running your distro as my main media center which runs a few other apps like sonarr and couchpotato on my Pi3

I am happy to make the module available for such use cases, but it may not necessarily benefit everyone. Your use case is more specific, and I believe people will always be able to optimise better for their application. However, if there's something we can do that you believe will benefit all users, I'm more than happy to take a look.

**Comment by johnfocker** [comment]
Hi! Thank you for taking the time to look into that as I understand your point of view and my use case is probably not the most wide spread but I chose to install OSMC in the first place instead of something like LibreELEC because OSMC gives me the best Kodi experience on top of a full featured Raspbian so I can get my media center up and running very quickly but still I can add features as time passes just like I did when adding Transmission-daemon, CouchPotato and Sonarr.

I would tipically use OSMC (Kodi) interface on my tv to display content while running the download servers which save the content on a big (slow) external hard drive connected through one of the (slow) USB2 interfaces of the RPi3.

With my current settings, memory usage is regurlarly above 70% which decreases in turn the size of cache.

My objective here is to make my OSMC instance as responsive as possible.

Tweaking the niceness and the ioniceness of the kodi process gives good results but I'd like to try and go further by adding a zswap and increasing the swappiness to give the more room possible to the buffers and cache.

I'm currently recompiling your kernel with the added options cited above but I was wondering why you would remove a module like zram which seems to me like a good idea to keep for low memory devices especially since it won't be loaded and used unless you explicitly tell your system to.

tl;dr could you keep modules like zram, zswap... in your kernel for geeks like me since it won't make a difference for "normal" users please?
I'll be sure to keep you posted with benchmarks and configurations in the case they prove to be good if you're interested.

**Comment by samnazarko** [comment]
>Tweaking the niceness and the ioniceness of the kodi process gives good results but I'd like to try and go further by adding a zswap and increasing the swappiness to give the more room possible to the buffers and cache.

I'm still interested in which tweaks you made. Did you test them on devices with lower memory, such as the Raspberry Pi Model B?

>My objective here is to make my OSMC instance as responsive as possible.

Certainly mine too, but we have most of the low hanging fruit picked. This is likely a workload specific optimisation that will not benefit other users. However I'm willing to give it some testing. Let me know what you did.

>I was wondering why you would remove a module like zram which seems to me like a good idea to keep for low memory devices especially since it won't be loaded and used unless you explicitly tell your system to.

It's more that it was not added, as we haven't seen a use for it, rather than 'removed'. 

I believe your workload is IO bound, I'm not sure that increasing swappiness will help (we have no swap by default for Pi 3). I could see some potential in renicing Kodi. If memory serves we already allow Kodi to do this when necessary. However your workload is quite specific: most people using OSMC have a lot of free memory spare and are not running significantly burdensome IO workloads. It's plausible we can optimise for your workload, but there will always be improvements that can be made manually as we cannot cater for all scenarios. 

I can certainly add the modules. Let me know what CONFIG_ options are needed.

Cheers

**Comment by johnfocker** [comment]
> I'm still interested in which tweaks you made. Did you test them on devices with lower memory, such as the Raspberry Pi Model B?

Sorry I don't have any other device than that one rbp3 I got so no but basically I added those 2 options to the systemd unit file `/lib/systemd/system/mediacenter.service`:
<pre><code>
[Unit]
Description = media center application
After = remote-fs.target network-online.target mysql.service
Wants = network-online.target

[Service]
Type = simple
TimeoutStopSec = 20
ExecStart = /usr/bin/mediacenter
ExecStopPost = /usr/bin/mediacenter stop
Restart = on-abort
<b>Nice=-10
IOSchedulingClass=1</b>

[Install]
WantedBy = multi-user.target
</code></pre>
Which I guess is bad practice in case this file gets updated, my changes will disappear.
Maybe a good approach is the one adopted by the raspbian packager for kodi who provides a settings file `/etc/default/kodi` so you can set the niceness of the kodi process (which they set to -5 by default). I didn't find that file for the OSMC version of Kodi?

This sets kodi's niceness to -10 and its ioniceness to class 1 (realtime). Those settings work very well for me but while setting the niceness to -10 is probably safe, I'm not sure wether settting the IO niceness to realtime would be advisable for general deployment as it might starve other critical processes in IO as ionice man page suggests but then again your system is supposed to only run OSMC's Kodi so it may not be a problem for you either. [The man](http://man7.org/linux/man-pages/man2/ioprio_set.2.html) says by default the IO niceness follows the process niceness but it's not what I observed so to be sure I specify the ioniceness to class 1.
Anyway those just two settings really improved my kodi process responsiveness (while I also reduced my other services -transmission/couchpotato/sonarr- priorities).

> It's more that it was not added, as we haven't seen a use for it, rather than 'removed'. 

My bad, I'd guessed you guys built upon the raspbian config and I knew it contained the zram module so I was looking for a performance related reason to effectively remove the module but couldn't find any.

> I believe your workload is IO bound, I'm not sure that increasing swappiness will help (we have no swap by default for Pi 3).

I didn't activate any swap file or partition on my sdcard either, yes that would probably be a disaster considering my workload :)
I use the zram devices as swap so whenever my system sends pages to swap, it actually just compresses them. It still is in RAM, just compressed.
As you mentioned, my workload is IO bound and that's why from my understanding of memory management I want my system to compress less used pages of memory instead of reducing the cache area before swaping (which isn't much slower than remaining uncompressed, since all my swap is actually just lz4 zram devices).
So I put in `/etc/sysctl.d/osmc.conf`:
```
vm.swappiness=100
vm.vfs_cache_pressure=1
```

If you're interested I used the script from the [Debian wiki](https://wiki.debian.org/ZRam) to activate my zram devices and I use `FRACTION=100` instead of 75 (no pb here, the lz4 compression algo promises a 1/3 ratio)
The default comp algorithm is lzo so you also need to declare something like `echo lz4 > /sys/block/zram$i/comp_algorithm` before the line `echo $SIZE > /sys/block/zram$i/disksize`
lzo and lz4 are very close in performance and compression ratio [but lz4 is about 3x faster to decompress](http://catchchallenger.first-world.info/wiki/Quick_Benchmark:_Gzip_vs_Bzip2_vs_LZMA_vs_XZ_vs_LZ4_vs_LZO).

As for the needed kernel modules needed:
`+CONFIG_ZSMALLOC=m` a zram dependancy
`+CONFIG_PGTABLE_MAPPING=y` [zsmalloc optimisation for ARM type architectures](https://cateee.net/lkddb/web-lkddb/PGTABLE_MAPPING.html)
`+CONFIG_ZSMALLOC_STAT=n` because I don't think we need that
`+CONFIG_ZRAM=m`
`+CONFIG_ZRAM_LZ4_COMPRESS=y`

Thanks again!

**Comment by samnazarko** [comment]
Can you adjust that PR for all platforms and squash? I'd like to keep the kernel configurations similar across all devices.

I'm not sure about statically renicing Kodi in that way yet, and it introduces a variable that most users will be unaware of. 

>vm.swappiness=100
>vm.vfs_cache_pressure=1

It seems like what you really need is a device with more available memory.

I'm happy to make the modules available.  I'll need to look in to CONFIG_PGTABLE_MAPPING=y in some detail too.

**Comment by johnfocker** [comment]
Sure, I was thinking about it but I first need to check the different architectures OSMC supports in order to set `CONFIG_PGTABLE_MAPPING` accordingly.
I'll look for benchmarks to confirm what I read in the doc about this parameter also.

Cheers

**Comment by issame** [comment]
Has this been implemented on current OSMC releases?

**Comment by samnazarko** [comment]
#338 is open, but I want a PR that makes this change for all platforms for consistency.

**Comment by johnfocker** [comment]
Sorry I had completely forgot about that PR and I just updated it.
Let me know if it looks good.

Congratulations for the vero 4k btw, it looks awesome.

**Comment by samnazarko** [comment]
Thanks @johnfocker -- did you edit this by hand or edit the .config files with make menuconfig?

**Comment by johnfocker** [comment]
By hand actually since the only dependency is zsmalloc, is this a problem?

**Comment by samnazarko** [comment]
Should be OK. I'll look at this and let you know.

**Comment by johnfocker** [comment]
Thanks :)

**Comment by samnazarko** [comment]
Support has been added, so closing.
